libbpf-tools/cpufreq.c

```diff
@@ -0,0 +1,239 @@
+				"%d\n", i);
+			links[i] = NULL;
+			close(fd);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+}
+
+static int init_freqs_hmz(__u32 *freqs_mhz, int nr_cpus)
+{
+	char *key = "cpu MHz", *p, line[32];
+	int len = strlen(key), freq_hmz, i;
+	FILE *f;
+
+	f = fopen("/proc/cpuinfo", "r");
+	if (!f)
+		return -1;
+
+	while (fgets(line, sizeof(line), f)) {
+		if (strncmp(line, key, len))
+			continue;
+
+		for (p = line + len; isspace(*p); p++);
+		if (*p != ':')
+			goto err_out;
+		for (++p; isspace(*p); p++);
+		if (!*p)
+			goto err_out;
+		freq_hmz = atoi(p);
+		for (i = 0; i < nr_cpus; i++)
+			freqs_mhz[i] = freq_hmz;
+
+		fclose(f);
+		return 0;
+	}
+
+err_out:
+	fclose(f);
+	return -1;
+}
+
+static void print_linear_hists(struct bpf_map *hists,
+			struct cpufreq_bpf__bss *bss)
+{
+	struct hkey lookup_key = {}, next_key;
+	int err, fd = bpf_map__fd(hists);
+	struct hist hist;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return;
+		}
+		print_linear_hist(hist.slots, MAX_SLOTS, 0, HIST_STEP_SIZE,
+				next_key.comm);
+		printf("\n");
+		lookup_key = next_key;
+	}
+
+	printf("\n");
+	print_linear_hist(bss->syswide.slots, MAX_SLOTS, 0, HIST_STEP_SIZE,
+			"syswide");
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct bpf_link **links = NULL;
+	struct cpufreq_bpf *obj;
+	int err, i;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	nr_cpus = libbpf_num_possible_cpus();
+	if (nr_cpus > MAX_CPU_NR) {
+		fprintf(stderr, "The number of cpu cores is too much, please "
+			"increase MAX_CPU_NR's value and recompile");
+		return 1;
+	}
+
+	obj = cpufreq_bpf__open_and_load();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF object\n");
+		return 1;
+	}
+
+	if (init_freqs_hmz(obj->bss->freqs_mhz, nr_cpus)) {
+		fprintf(stderr, "failed to init freqs\n");
+		goto cleanup;
+	}
+
+	links = calloc(nr_cpus, sizeof(*links));
+	if (!links) {
+		fprintf(stderr, "failed to alloc links\n");
+		goto cleanup;
+	}
+
+	if (open_and_attach_perf_event(FREQ, obj->progs.do_sample, links))
+		goto cleanup;
+
+	err = cpufreq_bpf__attach(obj);
+	if (err) {
+		fprintf(stderr, "failed to attach BPF programs\n");
+		goto cleanup;
+	}
+
+	printf("Sampling CPU freq system-wide & by process. Ctrl-C to end.\n");
+
+	signal(SIGINT, sig_handler);
+
+	sleep(env.duration);
```

Comment:
It's nice to have a comment here mentioning that we'll get sleep interrupted when someone presses Ctrl-C (which will be "handled" with noop by sig_handler). It's not obvious on initial reading.

Reply:
Sure, thanks!

Reply:
Fixed!